Include/internal/pycore_lock.h

```diff
@@ -93,6 +93,16 @@ PyAPI_FUNC(void) PyEvent_Wait(PyEvent *evt);
 PyAPI_FUNC(int)
 PyEvent_WaitTimed(PyEvent *evt, PyTime_t timeout_ns, int detach);
 
+// A one-time event notification with reference counting
+typedef struct _PyEventRc {
+    PyEvent event;
+    Py_ssize_t refcount;
+} _PyEventRc;
+
+extern _PyEventRc *_PyEventRc_New(void);
+extern void _PyEventRc_Incref(_PyEventRc *erc);
+extern void _PyEventRc_Decref(_PyEventRc *erc);
```

Comment:
This PR would be a lot simpler if we could avoid adding this new refcounted event type.  Could we store a reference to the handle on `PyThreadState` instead of a reference to the event?

Reply:
One alternative would be to trigger the event for all daemon threads when "shutting down" threads.  Perhaps we could use a separate linked list on the module to track daemon threads.

Reply:
For what it's worth, I would also lean towards triggering the event for daemon threads when shutting down.

1) We already have a linked list of `ThreadHandles` per-runtime (for after-fork handling). We can do another similar per-interpreter list.
2) I think we could do the notification somewhere around here: https://github.com/python/cpython/blob/a15a584bf3f94ea11ab9363548c8872251364000/Python/pylifecycle.c#L2041-L2043

Reply:
> Could we store a reference to the handle on PyThreadState instead of a reference to the event?

Yep, see my reply to your earlier comment. I'll give that a go.

> One alternative would be to trigger the event for all daemon threads when "shutting down" threads.

I considered this, but feel a bit uncomfortable with the temporal gap between marking them as done and when the threads actually exit as part of finalization. Maybe it isn't so bad? It would mean we don't have to reintroduce anything into `PyThreadState`.

Reply:
Just saw Sam's comments. Triggering the event for daemon threads seems reasonable, I'll give that a go.